src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,542 @@
+        }
+
+        foreach ($builder->whereIns as $field => $value) {
+            $compound['filter'][] = ['in' => ['path' => $field, 'value' => $value]];
+        }
+
+        foreach ($builder->whereNotIns as $field => $value) {
+            $compound['mustNot'][] = ['in' => ['path' => $field, 'value' => $value]];
+        }
+
+        // Sort by field value only if specified
+        $sort = [];
+        foreach ($builder->orders as $order) {
+            $sort[$order['column']] = $order['direction'] === 'asc' ? 1 : -1;
+        }
+
+        $pipeline = [
+            [
+                '$search' => [
+                    'index' => self::INDEX_NAME,
+                    'compound' => $compound,
+                    'count' => ['type' => 'lowerBound'],
+                    ...($sort ? ['sort' => $sort] : []),
+                ],
+            ],
+            [
+                '$addFields' => [
+                    // Metadata field with the total count of documents
+                    '__count' => '$$SEARCH_META.count.lowerBound',
+                ],
+            ],
+        ];
+
+        if ($offset) {
+            $pipeline[] = ['$skip' => $offset];
+        }
+
+        if ($builder->limit) {
+            $pipeline[] = ['$limit' => $builder->limit];
+        }
+
+        $options = [
+            'typeMap' => ['root' => 'array', 'document' => 'array', 'array' => 'array'],
+        ];
+
+        return $collection->aggregate($pipeline, $options)->toArray();
+    }
+
+    /**
+     * Pluck and return the primary keys of the given results.
+     *
+     * @see Engine::mapIds()
+     *
+     * @param list<array|object> $results
+     */
+    #[Override]
+    public function mapIds($results): Collection
+    {
+        assert(is_array($results), new TypeError(sprintf('Argument #1 ($results) must be of type array, %s given', get_debug_type($results))));
+
+        return new Collection(array_column($results, '_id'));
+    }
+
+    /**
+     * Map the given results to instances of the given model.
+     *
+     * @see Engine::map()
+     *
+     * @param array $results
+     * @param Model $model
+     */
+    #[Override]
+    public function map(Builder $builder, $results, $model): Collection
+    {
+        assert(is_array($results), new TypeError(sprintf('Argument #2 ($results) must be of type array, %s given', get_debug_type($results))));
+        assert($model instanceof Model, new TypeError(sprintf('Argument #3 ($model) must be of type %s, %s given', Model::class, get_debug_type($model))));
+
+        if (! $results) {
+            return $model->newCollection();
+        }
+
+        $objectIds = array_column($results, '_id');
+        $objectIdPositions = array_flip($objectIds);
+
+        return $model->getScoutModelsByIds(
+            $builder,
+            $objectIds,
+        )->filter(function ($model) use ($objectIdPositions) {
+            return array_key_exists($model->getScoutKey(), $objectIdPositions);
```

Comment:
It's not clear to me why this `filter()` step is required, but I see that all of the other engines in laravel/scout do the same.

Reply:
It should not be necessary. I think that's a double-check if the query returned results that are not in the `$objectIds` list. Maybe for the `queryCallback` that can be configured.